File: 백준(python)/baekjoon_2839.py
# https://www.acmicpc.net/problem/2839

import sys

N = int(sys.stdin.readline().strip())

fiveCount = 0
threeCount = 0
# 정확하게 N킬로그램을 만드는 답들의 리스트
resBasket = []

# 배달할 설탕의 3, 5의 모든 개수 조합 찾기
for count in range((N // 3) + 1):
    threeCount = count
    fiveCount = (N - 3 * count) // 5
    
    if fiveCount * 5 + threeCount * 3 == N:
        resBasket.append(fiveCount + threeCount)

# 리스트에 아무것도 존재하지 않을 경우, 정확하게 N킬로그램에 부합하는 값이 없기때문에
# -1 반환 
if not resBasket:
    print(-1)
else:
    print(min(resBasket))

File: 백준/baekjoon_2839.java
package baekjoon;

//https://www.acmicpc.net/problem/2839

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class baekjoon_2839 {
	
	public static void main(String[] args) throws NumberFormatException, IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		
		// 특정 크기를 단정할 수 없기때문에
		// 메모리가 허용하는 한 추가 가능한 List 사용
		List resBasket = new ArrayList();
		int fiveCount = 0;
		int threeCount = 0;
		
		for(int i = 0; i <= (N/5); i++) {
			fiveCount = i;
			threeCount = (N - 5 * i) / 3;

			if(fiveCount * 5 + threeCount * 3 == N) {
				resBasket.add(fiveCount + threeCount);
			}
		}
		
		// 리스트가 비어있을 경우 조건에 부합하는 답이 없음을 의미하기때문에 -1반환
		if(resBasket.isEmpty()) {
			System.out.println(-1);
		}
		else {
			// List내의 가장 작은 값 출력
			System.out.println(Collections.min(resBasket));
		}
	}
}
